CMakeLists.txt

```diff
@@ -220,51 +221,87 @@ if(BUILD_PYTHON)
+# This is just a CMake quirk; an OBJECT library is a better way to implement a
+# "private" (not installed) static library.
+add_library(codegen_internal OBJECT ${NVFUSER_SRCS})
 
 if(NOT MSVC)
-  target_compile_options(${NVFUSER_CODEGEN} PRIVATE
+  target_compile_options(codegen_internal PRIVATE
     -Wall -Wno-unused-function
     -Werror
   )
 endif()
 
-target_compile_definitions(${NVFUSER_CODEGEN} PRIVATE "-DTORCH_CUDA_BUILD_MAIN_LIB")
-target_include_directories(${NVFUSER_CODEGEN} SYSTEM PRIVATE
-  ${CMAKE_SOURCE_DIR}/third_party/flatbuffers/include
+target_compile_definitions(codegen_internal PRIVATE "-DTORCH_CUDA_BUILD_MAIN_LIB")
+target_include_directories(codegen_internal SYSTEM PUBLIC
   ${CMAKE_SOURCE_DIR}/third_party/gloo # TODO: guard this on usage
+  ${CMAKE_SOURCE_DIR}/third_party/flatbuffers/include
+	PRIVATE
   ${CUDA_TOOLKIT_ROOT_DIR}/extras/CUPTI/include
   ${CUDA_INCLUDE_DIRS}
 )
-target_include_directories(${NVFUSER_CODEGEN} PUBLIC
+target_include_directories(codegen_internal PUBLIC
   "$<BUILD_INTERFACE:${NVFUSER_SRCS_DIR}>"
   "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/nvfuser>"
 )
-set_property(TARGET ${NVFUSER_CODEGEN} PROPERTY CXX_STANDARD_REQUIRED ON)
-set_property(TARGET ${NVFUSER_CODEGEN} PROPERTY CXX_STANDARD ${NVFUSER_CPP_STANDARD})
+set_target_properties(codegen_internal PROPERTIES
+  C_STANDARD 17
+  CUDA_STANDARD 17 # CUDA 11 does not support C++20
+  CXX_STANDARD ${NVFUSER_CPP_STANDARD}
+  CXX_STANDARD_REQUIRED ON
+  CXX_VISIBILITY_PRESET hidden
+  # this is to find pip installed nvrtc/nvtx .so
+  INSTALL_RPATH
+  "$ORIGIN/../../nvidia/cuda_runtime/lib:$ORIGIN/../../nvidia/cuda_nvrtc/lib:$ORIGIN/../../nvidia/nvtx/lib:$ORIGIN/../../nvidia/cuda_cupti/lib:$ORIGIN/../../torch/lib"
+  POSITION_INDEPENDENT_CODE Yes
+  VISIBILITY_INLINES_HIDDEN Yes
+)
 
 # Ensure we don't link against libcuda; we'll dlopen it ourselves.
 list(FILTER TORCH_LIBRARIES EXCLUDE REGEX "libcuda\.so")
-target_link_libraries(${NVFUSER_CODEGEN} PRIVATE
-  flatbuffers
+target_link_libraries(codegen_internal PUBLIC
   dynamic_type
-  ${CUDA_NVRTC_LIB}
-  ${LIBNVTOOLSEXT}
   ${LIBCUPTI}
   ${TORCH_LIBRARIES}
   dl
 )
 
 if(NVFUSER_BUILD_WITH_ASAN)
-  add_compile_options(-fsanitize=address)
-  add_link_options(-fsanitize=address)
+  target_compile_options(codegen_internal PRIVATE -fsanitize=address)
+  target_link_options(codegen_internal PRIVATE -fsanitize=address)
 endif()
 
-# this is to find pip installed nvrtc/nvtx .so
-set_target_properties(${NVFUSER_CODEGEN} PROPERTIES INSTALL_RPATH
-  "$ORIGIN/../../nvidia/cuda_runtime/lib:$ORIGIN/../../nvidia/cuda_nvrtc/lib:$ORIGIN/../../nvidia/nvtx/lib:$ORIGIN/../../nvidia/cuda_cupti/lib:$ORIGIN/../../torch/lib")
-install(TARGETS ${NVFUSER_CODEGEN} EXPORT NvfuserTargets DESTINATION lib)
+add_library(nvfuser_codegen SHARED $<TARGET_OBJECTS:codegen_internal>)
+target_link_libraries(nvfuser_codegen PRIVATE
+  flatbuffers
+  ${CUDA_NVRTC_LIB}
+  ${LIBNVTOOLSEXT}
+  ${LIBCUPTI}
+  ${TORCH_LIBRARIES}
```

Comment:
I thought nvfuser_codegen would just inherit the dependency from the target. Do we need to have `LIBCUPT` and `TORCH_LIBRARIES` here, since they are already in `codegen_internal` as public.

Reply:
Yeah, I thought so too. I tried it though, it doesn't work. 🤷